lib/assignment_pipeline.rb

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+# Manages the current status of an assignment.
+class AssignmentPipeline
+  module AssignmentStatuses
+    IN_PROGRESS = 'in_progress'
+    READY_FOR_REVIEW = 'ready_for_review'
+    READY_FOR_MAINSPACE = 'ready_for_mainspace'
+  end
+
+  ASSIGNMENT_STATUS_PIPELINE = [
+    AssignmentStatuses::IN_PROGRESS,
+    AssignmentStatuses::READY_FOR_REVIEW,
+    AssignmentStatuses::READY_FOR_MAINSPACE
+  ]
+
+  module ReviewStatuses
+    PEER_REVIEWER_NEEDED = 'peer_reviewer_needed'
+    PEER_REVIEW_STARTED = 'peer_review_started'
+    PEER_REVIEW_COMPLETED = 'peer_review_completed'
+  end
+
+  REVIEW_STATUS_PIPELINE = [
+    ReviewStatuses::PEER_REVIEWER_NEEDED,
+    ReviewStatuses::PEER_REVIEW_STARTED,
+    ReviewStatuses::PEER_REVIEW_COMPLETED
+  ]
+
+  def initialize(flags:)
+    @flags = flags
+  end
+
+  def assignment_status
+    last_status(@flags[:assignment_status]) || ASSIGNMENT_STATUS_PIPELINE.first
+  end
+
+  def next_assignment_status
+    next_status(ASSIGNMENT_STATUS_PIPELINE, assignment_status)
+  end
+
+  def increment_assignment_status
+    @flags[:assignment_status] ||= []
+    @flags[:assignment_status] << [ next_assignment_status, Time.now ]
```

Comment:
This potentially allows any number of calls to `increment_assignment_status`, and it will fill the array with `[nil, <time>]` once it reaches the last status, I think.

If it's going to be a strictly linear status pipeline, I don't think there's a need to store every status change; just keeping the current status and its timestamp should be a good representation of the state.